File: benchmarks2d/convex_polygons2.rs
use rand::distr::{Distribution, StandardUniform};
use rand::{SeedableRng, rngs::StdRng};
use rapier_testbed2d::Testbed;
use rapier2d::prelude::*;

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut bodies = RigidBodySet::new();
    let mut colliders = ColliderSet::new();
    let impulse_joints = ImpulseJointSet::new();
    let multibody_joints = MultibodyJointSet::new();

    /*
     * Ground
     */
    let ground_size = 30.0;

    let rigid_body = RigidBodyBuilder::fixed();
    let handle = bodies.insert(rigid_body);
    let collider = ColliderBuilder::cuboid(ground_size, 1.2);
    colliders.insert_with_parent(collider, handle, &mut bodies);

    let rigid_body = RigidBodyBuilder::fixed()
        .rotation(std::f32::consts::FRAC_PI_2)
        .translation(vector![ground_size, ground_size * 2.0]);
    let handle = bodies.insert(rigid_body);
    let collider = ColliderBuilder::cuboid(ground_size * 2.0, 1.2);
    colliders.insert_with_parent(collider, handle, &mut bodies);

    let rigid_body = RigidBodyBuilder::fixed()
        .rotation(std::f32::consts::FRAC_PI_2)
        .translation(vector![-ground_size, ground_size * 2.0]);
    let handle = bodies.insert(rigid_body);
    let collider = ColliderBuilder::cuboid(ground_size * 2.0, 1.2);
    colliders.insert_with_parent(collider, handle, &mut bodies);

    /*
     * Create the convex polygons
     */
    let num = 26;
    let scale = 2.0;
    let border_rad = 0.0;

    let shift = border_rad * 2.0 + scale;
    let centerx = shift * (num as f32) / 2.0;
    let centery = shift / 2.0;

    let mut rng = StdRng::seed_from_u64(0);
    let distribution = StandardUniform;

    for i in 0..num {
        for j in 0usize..num * 5 {
            let x = i as f32 * shift - centerx;
            let y = j as f32 * shift * 2.0 + centery + 2.0;

            let rigid_body = RigidBodyBuilder::dynamic().translation(vector![x, y]);
            let handle = bodies.insert(rigid_body);

            let mut points = Vec::new();

            for _ in 0..10 {
                let pt: Point<f32> = distribution.sample(&mut rng);
                points.push(pt * scale);
            }

            let collider = ColliderBuilder::convex_hull(&points).unwrap();
            colliders.insert_with_parent(collider, handle, &mut bodies);
        }
    }

    /*
     * Set up the testbed.
     */
    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints);
    testbed.look_at(point![0.0, 50.0], 10.0);
}

File: python/rapier-testbed/rapier_testbed/examples2/stress_tests/convex_polygons2.py
"""Port of examples2d/stress_tests/convex_polygons2.rs."""
from __future__ import annotations

import math
import random

import numpy as np

import rapier2d as rp
from ..._registry import register

CATEGORY = "Stress Tests"
NAME = "Convex polygons"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    ground_size = 30.0

    handle = bodies.insert(rp.RigidBody.fixed())
    colliders.insert_with_parent(rp.Collider.cuboid(ground_size, 1.2), handle, bodies)

    handle = bodies.insert(
        rp.RigidBody.fixed()
        .rotation(math.pi / 2.0)
        .translation((ground_size, ground_size * 2.0))
    )
    colliders.insert_with_parent(
        rp.Collider.cuboid(ground_size * 2.0, 1.2), handle, bodies
    )

    handle = bodies.insert(
        rp.RigidBody.fixed()
        .rotation(math.pi / 2.0)
        .translation((-ground_size, ground_size * 2.0))
    )
    colliders.insert_with_parent(
        rp.Collider.cuboid(ground_size * 2.0, 1.2), handle, bodies
    )

    num = 26
    scale = 2.0
    shift = scale
    centerx = shift * num / 2.0
    centery = shift / 2.0

    rng = random.Random(0)

    for i in range(num):
        for j in range(num * 5):
            x = i * shift - centerx
            y = j * shift * 2.0 + centery + 2.0
            handle = bodies.insert(rp.RigidBody.dynamic().translation((x, y)))
            pts = np.array(
                [
                    [rng.random() * scale, rng.random() * scale]
                    for _ in range(10)
                ],
                dtype=np.float32,
            )
            colliders.insert_with_parent(rp.Collider.convex_hull(pts), handle, bodies)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.set_camera_2d(center=(0.0, 50.0), zoom=10.0)


register(CATEGORY, NAME, init_world, dim=2)

if __name__ == "__main__":
    from ..._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
